core/lib/zksync_core/src/reorg_detector/mod.rs

```diff
@@ -109,25 +140,53 @@ impl ReorgDetector {
     async fn is_legally_ahead_of_main_node(
         &self,
         sealed_l1_batch_number: L1BatchNumber,
+        sealed_miniblock_number: MiniblockNumber,
     ) -> RpcResult<bool> {
         // We must know the latest batch on the main node *before* we ask it for a root hash
         // to prevent a race condition (asked for root hash, batch sealed on main node, we've got
         // inconsistent results).
         let last_main_node_l1_batch = self.client.get_l1_batch_number().await?;
+        let last_main_node_miniblock = self.client.get_block_number().await?;
+
         let main_node_l1_batch_root_hash = self
             .client
             .get_l1_batch_details(sealed_l1_batch_number)
             .await?
             .and_then(|b| b.base.root_hash);
 
+        let main_node_miniblock_hash = self
+            .client
+            .get_block_by_number(last_main_node_miniblock.into(), false)
+            .await?
+            .unwrap()
+            .hash;
+
+        let local_miniblock_hash = self
+            .pool
+            .access_storage()
+            .await
+            .unwrap()
+            .blocks_dal()
+            .get_miniblock_header(sealed_miniblock_number)
+            .await
+            .unwrap()
+            .unwrap()
+            .hash;
+
         let en_ahead_for = sealed_l1_batch_number
             .0
             .checked_sub(last_main_node_l1_batch.as_u32());
+
         // Theoretically it's possible that the EN would not only calculate the root hash, but also seal the batch
         // quicker than the main node. So, we allow us to be at most one batch ahead of the main node.
         // If the gap is bigger, it's certainly a reorg.
         // Allowing the gap is safe: if reorg has happened, it'll be detected anyway in the future iterations.
-        Ok(main_node_l1_batch_root_hash.is_none() && en_ahead_for <= Some(1))
+        Ok(
+            sealed_miniblock_number.0 as u64 >= last_main_node_miniblock.as_u64()
+                && main_node_miniblock_hash == local_miniblock_hash
+                && main_node_l1_batch_root_hash.is_none()
+                && en_ahead_for <= Some(1),
+        )
```

Comment:
I think we can leave the only check here: `Ok(main_node_miniblock_hash.is_none() && en_ahead_for_miniblocks <= Some(1))`.

Reply:
Can you elaborate a bit? Bc to me this is not obvious bc why would the miniblock hash be none in this case for example?

Reply:
I mean EN can only be legally ahead for 1 miniblock, at least that's my understanding, @Deniallugo please confirm. Previously we used a weaker rule "EN can only be legally ahead for 1 L1 batch" but we can make it stricter as reorg detector is now aware of miniblocks.

Reply:
No, it can be legally ahead 1 batch, which what the current code suggests.

Reply:
I think that you can generally speaking be legally ahead for the arbitrary amounts of miniblocks in the case main node was reverted.

Reply:
AFAIU meaning of "legally" in this case assumes no reverts

Reply:
how can that happen without any rollbacks then even theoretically since EN cannot produce any of its own blocks?

Reply:
Good question, one thing is DB replicas can have different lags. So we can fetch some miniblock from one API server, process it, and then reorg detector will query another API server which is lagging. Most of the time lag is 0, but we don't want ENs to panic because of slight differences between API responses

Reply:
Ok, so essentially there is no guarantee that this lag will be only 1 block, right?

Reply:
Yes, leaving 1 batch lag allowed is more safe

Reply:
I suspect that at the end of the day simply being ahead whatever amounts of mibiblocks/batches is not a reorg scenario. We probably should be triggering the reorg only in the event when we see wrong hashes at the same height.

Reply:
Maybe, lagging behind too much should be catched somewhere IMO, but I'm not sure that reorg detector is a right place for it

Reply:
behind is completely another question, yes, talking only about ahead here

Reply:
oh, sorry, I meant ahead

Reply:
right, this makes sense then. I've updated code to trigger reorgs only if we have detected mismatches in block hashes at the same height.